File: lib/s/trace-query.ts
import type { RuleName } from "./cek"

export type TraceQueryField = "rule" | "detail" | "l"

export type TraceQueryAst =
  | { kind: "and"; left: TraceQueryAst; right: TraceQueryAst }
  | { kind: "or"; left: TraceQueryAst; right: TraceQueryAst }
  | { kind: "term"; field: TraceQueryField | null; value: string }

export type TraceQueryParseResult =
  | { ok: true; ast: TraceQueryAst | null }
  | { ok: false; message: string; at: number }

type TokenKind = "word" | "string" | "lparen" | "rparen" | "and" | "or" | "eq"

interface Token {
  kind: TokenKind
  text: string
  at: number
}

const FIELD_NAMES = new Set<TraceQueryField>(["rule", "detail", "l"])

export function parseTraceQuery(input: string): TraceQueryParseResult {
  const tokenized = tokenize(input)
  if (!tokenized.ok) return tokenized

  const tokens = tokenized.tokens
  if (tokens.length === 0) return { ok: true, ast: null }

  let pos = 0

  function current(): Token | undefined {
    return tokens[pos]
  }

  function consume(): Token {
    return tokens[pos++]
  }

  function parseOr(): TraceQueryParseResult {
    const first = parseAnd()
    if (!first.ok || first.ast === null) return first

    let ast = first.ast
    while (current()?.kind === "or") {
      const op = consume()
      const right = parseAnd()
      if (!right.ok) return right
      if (right.ast === null) {
        return { ok: false, message: "expected term after ||", at: op.at + 2 }
      }
      ast = { kind: "or", left: ast, right: right.ast }
    }
    return { ok: true, ast }
  }

  function parseAnd(): TraceQueryParseResult {
    const first = parsePrimary()
    if (!first.ok || first.ast === null) return first

    let ast = first.ast
    while (current()?.kind === "and") {
      const op = consume()
      const right = parsePrimary()
      if (!right.ok) return right
      if (right.ast === null) {
        return { ok: false, message: "expected term after &&", at: op.at + 2 }
      }
      ast = { kind: "and", left: ast, right: right.ast }
    }
    return { ok: true, ast }
  }

  function parsePrimary(): TraceQueryParseResult {
    const tok = current()
    if (!tok) {
      return { ok: true, ast: null }
    }

    if (tok.kind === "lparen") {
      consume()
      const inner = parseOr()
      if (!inner.ok) return inner
      if (inner.ast === null) {
        return { ok: false, message: "expected term after (", at: tok.at + 1 }
      }
      const close = current()
      if (close?.kind !== "rparen") {
        return {
          ok: false,
          message: "expected )",
          at: close?.at ?? input.length,
        }
      }
      consume()
      return inner
    }

    if (tok.kind === "word" || tok.kind === "string") {
      return parseTerm()
    }

    const suffix =
      tok.kind === "rparen"
        ? "before )"
        : tok.kind === "and"
          ? "before &&"
          : "before ||"
    return { ok: false, message: `expected term ${suffix}`, at: tok.at }
  }

  function parseTerm(): TraceQueryParseResult {
    const tok = consume()
    const eq = current()

    if (tok.kind === "word" && eq?.kind === "eq") {
      const field = tok.text.toLowerCase()
      if (!isTraceQueryField(field)) {
        return { ok: false, message: `unknown field ${tok.text}`, at: tok.at }
      }

      consume()
      const value = current()
      if (!value || (value.kind !== "word" && value.kind !== "string")) {
        return {
          ok: false,
          message: `expected value after ${tok.text}=`,
          at: value?.at ?? input.length,
        }
      }
      consume()
      if (value.text.length === 0) {
        return {
          ok: false,
          message: `expected value after ${tok.text}=`,
          at: value.at,
        }
      }
      return { ok: true, ast: { kind: "term", field, value: value.text } }
    }

    if (tok.text.length === 0) {
      return { ok: false, message: "expected term", at: tok.at }
    }
    return { ok: true, ast: { kind: "term", field: null, value: tok.text } }
  }

  const parsed = parseOr()
  if (!parsed.ok) return parsed

  const extra = current()
  if (extra) {
    if (extra.kind === "rparen") {
      return { ok: false, message: "unexpected )", at: extra.at }
    }
    return { ok: false, message: `unexpected ${extra.text}`, at: extra.at }
  }

  return parsed
}

export function traceQueryMatches(
  ast: TraceQueryAst | null,
  row: {
    index: number
    rule?: RuleName
    detail?: string
    value?: string
    label: number
  }
): boolean {
  if (ast === null) return true

  switch (ast.kind) {
    case "and":
      return (
        traceQueryMatches(ast.left, row) && traceQueryMatches(ast.right, row)
      )
    case "or":
      return (
        traceQueryMatches(ast.left, row) || traceQueryMatches(ast.right, row)
      )
    case "term":
      return termMatches(ast, row)
  }
}

function termMatches(
  term: Extract<TraceQueryAst, { kind: "term" }>,
  row: {
    rule?: RuleName
    detail?: string
    value?: string
    label: number
  }
): boolean {
  const needle = normalize(term.value)
  const label = String(row.label)

  if (term.field === "rule") return normalize(row.rule ?? "").includes(needle)
  if (term.field === "detail")
    return normalize(row.detail ?? "").includes(needle)
  if (term.field === "l") return normalize(label).includes(needle)

  return normalize(
    [row.rule ?? "", row.detail ?? "", row.value ?? "", label].join(" ")
  ).includes(needle)
}

function normalize(value: string): string {
  return value.toLowerCase()
}

function isTraceQueryField(value: string): value is TraceQueryField {
  return FIELD_NAMES.has(value as TraceQueryField)
}

function tokenize(
  input: string
): { ok: true; tokens: Token[] } | { ok: false; message: string; at: number } {
  const tokens: Token[] = []
  let i = 0

  while (i < input.length) {
    const ch = input[i]

    if (/\s/.test(ch)) {
      i++
      continue
    }

    if (ch === "(") {
      tokens.push({ kind: "lparen", text: ch, at: i })
      i++
      continue
    }

    if (ch === ")") {
      tokens.push({ kind: "rparen", text: ch, at: i })
      i++
      continue
    }

    if (ch === "=") {
      tokens.push({ kind: "eq", text: ch, at: i })
      i++
      continue
    }

    if (ch === "&" || ch === "|") {
      const next = input[i + 1]
      if (ch === "&" && next === "&") {
        tokens.push({ kind: "and", text: "&&", at: i })
        i += 2
        continue
      }
      if (ch === "|" && next === "|") {
        tokens.push({ kind: "or", text: "||", at: i })
        i += 2
        continue
      }
      return { ok: false, message: `expected ${ch}${ch}`, at: i }
    }

    if (ch === '"') {
      const start = i
      i++
      let text = ""
      let closed = false
      while (i < input.length) {
        const q = input[i]
        if (q === '"') {
          i++
          tokens.push({ kind: "string", text, at: start })
          closed = true
          break
        }
        if (q === "\\" && i + 1 < input.length) {
          text += input[i + 1]
          i += 2
          continue
        }
        text += q
        i++
      }
      if (!closed) {
        return { ok: false, message: "unterminated string", at: start }
      }
      continue
    }

    const start = i
    let text = ""
    while (i < input.length) {
      const q = input[i]
      if (
        /\s/.test(q) ||
        q === "(" ||
        q === ")" ||
        q === "=" ||
        q === "&" ||
        q === "|"
      ) {
        break
      }
      if (q === '"') {
        return { ok: false, message: "unexpected string quote", at: i }
      }
      text += q
      i++
    }
    tokens.push({ kind: "word", text, at: start })
  }

  return { ok: true, tokens }
}

File: lib/s/traceQuery.ml
module I = TraceQueryGrammar.MenhirInterpreter
open TraceQueryGrammar
include TraceQueryData

let token_text = function
  | LPAREN -> "("
  | RPAREN -> ")"
  | AND -> "&&"
  | OR -> "||"
  | NOT -> "!"
  | EQ -> "="
  | NEQ -> "!="
  | GT -> ">"
  | GTE -> ">="
  | LT -> "<"
  | LTE -> "<="
  | EOF -> ""
  | WORD text -> text
  | STRING text -> text

let syntax_error current input_len =
  match current with
  | None -> { message = "unexpected end of input"; at = input_len }
  | Some (EOF, _, _) -> { message = "unexpected end of input"; at = input_len }
  | Some (token, at, _) -> { message = "unexpected " ^ token_text token; at }

let parseTraceQuery input =
  let lexbuf = Lexing.from_string input in
  let input_len = String.length input in
  let rec loop checkpoint current =
    match checkpoint with
    | I.InputNeeded _ ->
        begin try
          let token = TraceQueryLexer.token lexbuf in
          let at = lexbuf.lex_start_p.pos_cnum in
          let startp = lexbuf.lex_start_p in
          let endp = lexbuf.lex_curr_p in
          loop
            (I.offer checkpoint (token, startp, endp))
            (Some (token, at, endp))
        with
        | TraceQueryLexer.Lex_error error -> Stdlib.Error error
        | Query_error error -> Stdlib.Error error
        end
    | I.Shifting _ | I.AboutToReduce _ -> loop (I.resume checkpoint) current
    | I.Accepted ast -> Stdlib.Ok ast
    | I.HandlingError _ | I.Rejected -> (
        match current with
        | None ->
            Stdlib.Error { message = "unexpected end of input"; at = input_len }
        | Some (EOF, _, _) ->
            Stdlib.Error { message = "unexpected end of input"; at = input_len }
        | Some (token, at, _) ->
            Stdlib.Error { message = "unexpected " ^ token_text token; at })
  in
  try loop (TraceQueryGrammar.Incremental.query_eof lexbuf.lex_curr_p) None
  with Query_error error -> Stdlib.Error error

let normalize value = String.lowercase_ascii value

let string_contains haystack needle =
  let haystack_len = String.length haystack in
  let needle_len = String.length needle in
  if needle_len = 0 then true
  else
    let rec loop i =
      if i + needle_len > haystack_len then false
      else if String.sub haystack i needle_len = needle then true
      else loop (i + 1)
    in
    loop 0

let string_of_rule_name = function
  | Cek.LetExp -> "LetExp"
  | Cek.LetCall -> "LetCall"
  | Cek.Match -> "Match"
  | Cek.Return -> "Return"
  | Cek.LetTag -> "LetTag"

let label_matches (term : term_payload) label =
  match term.op with
  | Eq -> string_of_int label = term.value
  | Gt ->
      begin match parse_integer_literal term.value with
      | Some value -> label > value
      | None -> false
      end
  | Gte ->
      begin match parse_integer_literal term.value with
      | Some value -> label >= value
      | None -> false
      end
  | Lt ->
      begin match parse_integer_literal term.value with
      | Some value -> label < value
      | None -> false
      end
  | Lte ->
      begin match parse_integer_literal term.value with
      | Some value -> label <= value
      | None -> false
      end

let term_matches (term : term_payload) (row : row) =
  let needle = normalize term.value in
  let label = string_of_int row.label in
  match term.field with
  | Some Rule ->
      let rule =
        match row.rule with
        | Some rule -> normalize (string_of_rule_name rule)
        | None -> ""
      in
      rule = needle
  | Some Detail ->
      let detail =
        match row.detail with Some detail -> normalize detail | None -> ""
      in
      string_contains detail needle
  | Some L -> label_matches term row.label
  | None ->
      let haystack =
        String.concat " "
          [
            (match row.rule with
            | Some rule -> string_of_rule_name rule
            | None -> "");
            (match row.detail with Some detail -> detail | None -> "");
            (match row.value with Some value -> value | None -> "");
            label;
          ]
        |> normalize
      in
      string_contains haystack needle

let rec traceQueryMatches ast row =
  match ast with
  | None -> true
  | Some (And { left; right }) ->
      traceQueryMatches (Some left) row && traceQueryMatches (Some right) row
  | Some (Or { left; right }) ->
      traceQueryMatches (Some left) row || traceQueryMatches (Some right) row
  | Some (Not expr) -> not (traceQueryMatches (Some expr) row)
  | Some (Term term) -> term_matches term row
